paddle/framework/attribute_reader.h

```diff
@@ -0,0 +1,193 @@
+  constexpr bool is_vec = std::is_same<T, std::vector<int>>::value ||
+                          std::is_same<T, std::vector<float>>::value ||
+                          std::is_same<T, std::vector<std::string>>::value;
+  return details::ContainsImpl<T, is_vec>()(attrs_, name);
+}
+
+#define ATTR_READER_ISTYPE_IMPL(T, CASE)                              \
+  namespace details {                                                 \
+  template <>                                                         \
+  inline bool IsType<T>(const ::paddle::framework::Attribute* attr) { \
+    return attr->value_case() == CASE;                                \
+  }                                                                   \
+  }
+
+ATTR_READER_ISTYPE_IMPL(int, ::paddle::framework::Attribute::kI);
+ATTR_READER_ISTYPE_IMPL(float, ::paddle::framework::Attribute::kF);
+ATTR_READER_ISTYPE_IMPL(std::string, ::paddle::framework::Attribute::kS);
+
+#undef ATTR_READER_ISTYPE_IMPL
+
+/// Implementation of Get
+namespace details {
+template <typename T>
+inline T GetValue(const ::paddle::framework::Attribute* attr);
+}
+
+template <typename T>
+T AttributeReader::Get(const std::string& name) const {
+  auto attr = details::GetField(attrs_, name);
+  PADDLE_ENFORCE(attr != nullptr, "Attribute %s not found", name);
+  PADDLE_ENFORCE(details::IsType<T>(attr),
+                 "Attribute type mismatch. Expected %s", typeid(T).name());
+  return details::GetValue<T>(attr);
+}
+
+#define ATTR_READER_GETVALUE_IMPL(T, FIELD)                          \
+  namespace details {                                                \
+  template <>                                                        \
+  inline T GetValue<T>(const ::paddle::framework::Attribute* attr) { \
+    return attr->FIELD();                                            \
+  }                                                                  \
+  }
+
+ATTR_READER_GETVALUE_IMPL(int, i);
+ATTR_READER_GETVALUE_IMPL(float, f);
+ATTR_READER_GETVALUE_IMPL(std::string, s);
+
+#undef ATTR_READER_GETVALUE_IMPL
+
+/// Implementation of GetArray
+#define ATTR_GETARRAY_IMPL(T, FIELD)                                     \
+  template <>                                                            \
+  void AttributeReader::GetArray<T>(const std::string& name,             \
+                                    std::vector<T>* vec) const {         \
+    PADDLE_ENFORCE(vec->empty(), "Input vector should be empty");        \
+    auto attr = details::GetField(attrs_, name);                         \
+    PADDLE_ENFORCE(attr != nullptr, "Attribute %s not found", name);     \
+    PADDLE_ENFORCE(attr->has_list(), "Attribute %s is not array", name); \
+    auto& field = attr->list().FIELD();                                  \
+    vec->reserve(field.size());                                          \
+    std::copy(field.begin(), field.end(), std::back_inserter(*vec));     \
+  }
+
+ATTR_GETARRAY_IMPL(int, ints);
```

Comment:
to be with the same name style with `ATTR_READER_GETVALUE_IMPL` and `ATTR_READER_ISTYPE_IMPL`

ATTR_GETARRAY_IMPL  =>  ATTR_READER_GETARRAY_IMPL